spec/member_holding_file_spec.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require "member_holding_file"
+require "custom_errors"
+
+RSpec.describe MemberHoldingFile do
+  let(:test_data) { __dir__ + "/../testdata" }
+
+  # Vars for tests expected to pass, prefix "ok_"
+  let(:ok_mem) { "umich" }
+  let(:ok_fn)  { "umich_mono_full_20201230_corrected.tsv" }
+  let(:ok_mhf) { described_class.new(ok_fn) }
+  let(:ok_itype) { "mono" }
+  let(:ok_col_map) { { "oclc"=>0, "local_id"=>1 } }
+
+  let(:ok_header_mhf) do
+    described_class.new(
+      "#{test_data}/haverford_mono_full_20200101_header.tsv"
+    )
+  end
+
+  # Vars for tests expected to fail, prefix "bad_"
+  let(:bad_mem) { "" }
+  let(:bad_fn)  { "FAIL ME PLEASE" }
+  let(:bad_mhf) { described_class.new(bad_fn) }
+
+  # each badness = !+
+  let(:bad_fn_1) { "!!!!!_mono_full_20201230.tsv" } # bad member_id
+  let(:bad_fn_2) { "umich_!!!!_full_20201230.tsv" } # bad item_type
+  let(:bad_fn_3) { "umich_mono_!!!!_20201230.tsv" } # bad update type
+  let(:bad_fn_4) { "umich_mono_full_!!!!!!!!.tsv" } # bad date
+  let(:bad_fn_5) { "umich_mono_full_20201230.!!!" } # bad extension
+
+  # file headers
+  let(:ok_hed_min) { ["oclc", "local_id"] }
+  let(:ok_hed_mul) { ["oclc", "local_id", "status", "condition", "enumchron", "govdoc"] }
+
+  it "finds member_id in filename" do
+    expect(ok_mhf.member_id_from_filename(ok_fn)).to eq(ok_mem)
+    expect(ok_mhf.member_id_from_filename).to eq(ok_mem)
+  end
+
+  it "raises error if no member_id in filename" do
+    expect { ok_mhf.member_id_from_filename("") }.to raise_error(FileNameError)
+    expect { ok_mhf.member_id_from_filename(nil) }.to raise_error(FileNameError)
+  end
+
+  it "finds item type in filename" do
+    expect(ok_mhf.item_type_from_filename).to eq(ok_itype)
+  end
+
+  it "raises error if no item_type in filename" do
+    expect { bad_mhf.item_type_from_filename }.to raise_error(FileNameError)
+  end
+
+  it "checks a whole filename" do
+    expect(ok_mhf.valid_filename?).to be(true)
+  end
+
+  it "Raises an error if it can't figure out member_id or item_type" do
```

Comment:
Consider inlining the `let`s you only use once, or moving them closer to the examples in which they're used, e.g.

```ruby
context "with bad filenames" do
    # each badness = !+
  let(:bad_fn_1) { "!!!!!_mono_full_20201230.tsv" } # bad member_id
  let(:bad_fn_2) { "umich_!!!!_full_20201230.tsv" } # bad item_type
  let(:bad_fn_3) { "umich_mono_!!!!_20201230.tsv" } # bad update type
  let(:bad_fn_4) { "umich_mono_full_!!!!!!!!.tsv" } # bad date
  let(:bad_fn_5) { "umich_mono_full_20201230.!!!" } # bad extension

  it "rejects ..."
end
```

Reply:
By "inlining let's" do you mean just straight up declaring a variable inside the test?